### ISSUE TITLE
GELF codec needs to support chunking

### ISSUE THREAD

**Issue by neuronull** [vectordotdev/vector#13292]
I'm noticing that GELF does have a chunking part of its protocol too, for multi-part messages when using UDP:

```
Prepend the following structure to your GELF message to make it chunked:

Chunked GELF magic bytes - 2 bytes: 0x1e 0x0f
Message ID - 8 bytes: Must be the same for every chunk of this message. Identifying the whole message and is used to reassemble the chunks later. Generate from millisecond timestamp + hostname for example.
Sequence number - 1 byte: The sequence number of this chunk. Starting at 0 and always less than the sequence count.
Sequence count - 1 byte: Total number of chunks this message has.
All chunks MUST arrive within 5 seconds or the server will discard all already arrived and still arriving chunks. A message MUST NOT consist of more than 128 chunks.
```

I think we can split this off into a follow up issue though and just start with basic decoding support of incoming messages.

_Originally posted by @jszwedko in https://github.com/vectordotdev/vector/pull/13288#issuecomment-1163483729_

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by neuronull** [comment]
This should be attempted after #4868 is closed.

**Comment by paluchnuggets** [comment]
Any update on this matter?

**Comment by angelrib** [comment]
@neuronull  

> I think we can split this off into a follow up issue though and just start with basic decoding support of incoming messages.

Does vector support chunking for output messages now? Or what you mean is start with the reassembling first?

Thank you

**Comment by neuronull** [comment]
Hi @angelrib , that comment actually came from @jszwedko  (https://github.com/vectordotdev/vector/pull/13288#issuecomment-1163483729)

So this (#13292) is the referred to follow-up issue to track chunking support.

I believe what Jesse meant is that we can probably just start with implementing chunking on the decoder side, and add the encoding side in a separate PR.

**Comment by angelrib** [comment]
Reviewing the code it looked like that, I just wanted to confirm it. Thank you very much!

**Comment by gernoteger** [comment]
Chunking imho is mandatory for UDP in order to be usable, since log messages would be limited to 8192 bytes otherwise, and fail depending on content.
Currently GELF is implemented as an encoding for various transports. Chunked GELF UDP has aspects of  a transport protocol, like dependencies on packets and a defined timeout for a sequence. Maybe this would need dedicated source and sink components.

**Comment by jorgehermo9** [comment]
Opened #20769 for addressing the decoding part. I'm currently working on it and will have it implemented soon. Just some details left

**Comment by jorgehermo9** [comment]
If anyone is tracking this, support for uncompressed chunked gelfs has been merged https://github.com/vectordotdev/vector/pull/20859.

I'll be addressing the decompression part, it can be tracked at #21153

**Comment by jorgehermo9** [comment]
Update: #21153 has been merged and released in Vector 0.43. 

Chunked gelf decoding is fully supported now. I may work in the encoding part to fully address this issue

**Comment by jszwedko** [comment]
Closing this issue since this has been implemented. Thank you for this large contribution @jorgehermo9 !

**Comment by jorgehermo9** [comment]
I thought this issue was for both encoding and decoding @jszwedko! The encoding part is not done yet

**Comment by jszwedko** [comment]
Ah, doh, thanks @jorgehermo9 . I actually forgot we had a GELF _encoder_. Reopening.

**Comment by aramperes** [comment]
I'm starting to look into this (GELF chunked encoding); and first thing I notice is that there is no current implementation of `encoding` or `framing` that splits an event into multiple `Bytes` buffers. 

Specifically looking at the UDP sink: it assumes that it is a single `BytesMut` buffer that is being encoded, then passed into a single datagram:

https://github.com/vectordotdev/vector/blob/031c2794dc96a21abedf0addf351f0ae59f7fda3/src/sinks/util/datagram.rs#L45-L50

This might require a custom `Encoder` trait method that has a signature like `fn encode_chunks(&mut self, item: Item) -> Result<Vec<bytes::Bytes>, Self::Error>`; every sink that supports `gelf` would need to be updated to implement the logic for sending multiple of their packet/datagram/payloads.

Alternatively, since chunking is a step that should happen _after_ encoding, it might be a new optional step in the pipeline: `chunking` / `Chunker` that defaults to a pass-through, but provides a trait for `fn create_chunks(&mut self, bytes::Bytes) -> Result<Vec<bytes::Bytes>, Self::Error>`.

Since UDP would be the sink that would most benefit from chunking, I would propose starting with only support it there as a first pass.

Does that make sense?

I have a proof-of-concept here: https://github.com/vectordotdev/vector/compare/master...aramperes:vector:feat-13292-gelf-chunking

**Comment by jorgehermo9** [comment]
Yep, that makes sense to me @aramperes . I tried to work on this before, and tokio codec's only supports to encode to a single `BytesMut` https://docs.rs/tokio-util/latest/tokio_util/codec/trait.Encoder.html. I ended up reaching similar conclusions to yours
